### PR TITLE
Make constructors for QgsVectorLayer and QgsRasterLayer more flexible, fix crs handling

### PR DESCRIPTION
**Pull request by nyalldawson**
...by moving extra arguments to new LayerOptions structs. This allows us to more easily add new layer constructor options without making the API cumbersome to use.

(This is making room to add an additional constructor option shortly)

### CODE REVIEW

**Comment by nyalldawson** [comment]
Second commit allows specifying a default layer CRS to use for vector  for when the actual provider CRS could not be determined.

If set, this CRS will be used when the actual CRS from  the data provider could not be determined, and instead of triggering the user's default unknown CRS handling method (e.g. opening the 'select layer CRS' dialog).
     
A pointer to an invalid QgsCoordinateReferenceSystem object can be used to indicate that the layer does not require a valid CRS and that no user prompts for the layer's CRS should be triggered.

If the crs argument is nullptr (the default), the default missing CRS handling method will be used intead.

This is designed to plug a major deficiency in the QGIS API, which forces hacky workarounds like:
https://github.com/qgis/QGIS/blob/master/src/core/processing/qgsprocessingutils.cpp#L135
https://github.com/qgis/QGIS/blob/master/src/core/processing/qgsprocessingutils.cpp#L170
https://github.com/qgis/QGIS/blob/master/src/gui/qgsbrowserdockwidget_p.cpp#L130

Basically, if a script needs to load a layer which has either a missing/undetectable CRS, it's only possible to avoid the appearance of the "select layer crs" dialog by a fragile hack of altering the projections setting key to disable this dialog, and then restore it after loading the layer. But this is horrible, and it's quite common that scripts will need to load up data formats with no available CRS information. (e.g. a script which downloads a netcdf file and then loads this into qgis).

TODO (if acceptable):
- tests
- similar changes for QgsRasterLayer

**Comment by luipir** [comment]
Could make sense to specify LayerOption struct as provider dependent? 
pro:
1) in this way constructor can be customized
2) In this moment source URI can specify too many different way tot manage the source... from a real URi to layer as query => customization options or parameters could move some of this not homogeneous parameters to a consistent API... btw I didn't though too much on this.
Cons:
1) no idea at what can be useful this customization other that to be flexible for  future constructor?
2) hide overloading with a struct can be unclean and error prone.

just my 2c

**Comment by nyalldawson** [comment]
@luipir

What do you mean exactly? An extra free-form string map?

**Comment by nyalldawson** [comment]
Well.. I guess that's the benefit of making this a struct. We can easily add new members during 3.x without breaking API.

**Comment by luipir** [comment]
I agree

**Comment by nyalldawson** [comment]
@wonder-sk I'd value your thoughts here. I now we've spoken in the past and you had plans for allowing undefined projection/"crs-less" layers. I've tried to handle that use case partly here.

**Comment by wonder-sk** [comment]
hi @nyalldawson great that you have looked into that!

My thoughts:
- I like the idea of LayerOptions for rasters and vectors
- my preference would be to actually move forced validation of CRS out of layer constructor. It could still stay inside QGIS app - when a layer is added to the project, the CRS validation would be run, but it would not be run for all layers including those that are never added to the current project and maybe do not even need a valid CRS. In a future release it would be nice to go even further and instead of doing forced validation we could load layers without valid CRS, just show a warning indicator next to layer's name, and let user set layer's CRS if needed. I have seen various times that novice users do not understand the purpose of CRS selection dialog that pops up when loading a new layer, ending up with wrong CRS assignments.

Somehow related thoughts :-)
- maybe worth moving layerName to LayerOptions as well? IIRC it will still get a name derived from URI if not specified - or it can be set anytime later. I always thought the layer constructors should have just two arguments: 1. uri, 2. provider - but never got to changing that. (Especially when writing pyqgis cookbook I thought why should we always pass layer name even if we don't care about it?)
- maybe readExtentFromXml could be renamed to something like "skipExtentDetection" ?
- maybe worth moving the boolean options into a flags enumeration? (no strong opinion here)

**Comment by nyalldawson** [comment]
Closing this. The important part for 3.0 landed, and the remainder (CRS handling for layers) requires a rework.
